File: src/services/message.service.ts
import feishuClient from '../bot/client';
import logger from '../utils/logger';

interface PermissionRequest {
  requestId: string;
  toolName: string;
  input: Record<string, any>;
}

class MessageService {
  // 给消息添加 emoji 反应，返回 reactionId
  async addReaction(messageId: string, emojiType: string): Promise<string | null> {
    try {
      const client = feishuClient.getClient();
      const res = await (client.im.messageReaction as any).create({
        path: { message_id: messageId },
        data: { reaction_type: { emoji_type: emojiType } },
      });
      return res?.data?.reaction_id ?? null;
    } catch (error) {
      logger.warn('Failed to add reaction', { messageId, emojiType, error });
      return null;
    }
  }

  // 移除消息的 emoji 反应
  async removeReaction(messageId: string, reactionId: string): Promise<void> {
    try {
      const client = feishuClient.getClient();
      await (client.im.messageReaction as any).delete({
        path: { message_id: messageId, reaction_id: reactionId },
      });
    } catch (error) {
      logger.warn('Failed to remove reaction', { messageId, reactionId, error });
    }
  }

  // 发送文本消息
  async sendTextMessage(chatId: string, text: string): Promise<void> {
    try {
      const client = feishuClient.getClient();

      await client.im.message.create({
        params: {
          receive_id_type: 'chat_id',
        },
        data: {
          receive_id: chatId,
          msg_type: 'text',
          content: JSON.stringify({ text }),
        },
      });

      logger.info('Text message sent', { chatId, text });
    } catch (error) {
      logger.error('Failed to send text message', { error, chatId });
      throw error;
    }
  }

  // 发送 Markdown 卡片消息
  async sendCardMessage(chatId: string, markdown: string): Promise<void> {
    try {
      const client = feishuClient.getClient();

      await client.im.message.create({
        params: {
          receive_id_type: 'chat_id',
        },
        data: {
          receive_id: chatId,
          msg_type: 'interactive',
          content: JSON.stringify({
            config: {
              wide_screen_mode: true,
            },
            elements: [
              {
                tag: 'markdown',
                content: markdown,
              },
            ],
          }),
        },
      });

      logger.info('Card message sent', { chatId, contentLength: markdown.length });
    } catch (error) {
      logger.error('Failed to send card message', { error, chatId });
      throw error;
    }
  }

  // 发送权限请求消息
  async sendPermissionRequest(chatId: string, request: PermissionRequest): Promise<void> {
    const inputStr = JSON.stringify(request.input, null, 2);
    const shortId = request.requestId.slice(0, 8);

    const card = {
      config: { wide_screen_mode: true },
      header: {
        title: { tag: 'plain_text', content: '🔐 权限请求' },
        template: 'orange',
      },
      elements: [
        {
          tag: 'div',
          text: { tag: 'lark_md', content: `**工具：** ${request.toolName}` },
        },
        {
          tag: 'markdown',
          content: `\`\`\`json\n${inputStr}\n\`\`\``,
        },
        { tag: 'hr' },
        {
          tag: 'div',
          text: {
            tag: 'lark_md',
            content: `回复：\n\`/approve ${shortId}\` - 批准\n\`/deny ${shortId}\` - 拒绝`,
          },
        },
      ],
    };

    try {
      const client = feishuClient.getClient();
      await client.im.message.create({
        params: { receive_id_type: 'chat_id' },
        data: {
          receive_id: chatId,
          msg_type: 'interactive',
          content: JSON.stringify(card),
        },
      });
      logger.info('Permission request sent', { chatId, requestId: request.requestId, shortId, toolName: request.toolName });
    } catch (error) {
      logger.error('Failed to send permission request', { chatId, error });
      throw error;
    }
  }
}

export default new MessageService();

File: src/services/message_service.py
"""消息发送服务"""
import lark_oapi as lark
from lark_oapi.api.im.v1 import *
import logging
import json
from src.bot.client import feishu_client

logger = logging.getLogger(__name__)


class MessageService:
    """消息发送服务"""

    def add_reaction(self, message_id: str, emoji_type: str) -> str | None:
        """添加消息表情反应"""
        try:
            client = feishu_client.get_client()
            request = CreateMessageReactionRequest.builder() \
                .message_id(message_id) \
                .request_body(CreateMessageReactionRequestBody.builder()
                    .reaction_type(Emoji.builder()
                        .emoji_type(emoji_type)
                        .build())
                    .build()) \
                .build()

            response = client.im.v1.message_reaction.create(request)

            if not response.success():
                logger.warning(f'Failed to add reaction: code={response.code}, msg={response.msg}', extra={
                    'message_id': message_id,
                    'code': response.code,
                    'msg': response.msg
                })
                return None

            return response.data.reaction_id

        except Exception as e:
            logger.warning(f'Failed to add reaction: {str(e)}', extra={
                'message_id': message_id,
                'error': str(e),
                'error_type': type(e).__name__
            })
            return None

    def remove_reaction(self, message_id: str, reaction_id: str):
        """移除消息表情反应"""
        try:
            client = feishu_client.get_client()
            request = DeleteMessageReactionRequest.builder() \
                .message_id(message_id) \
                .reaction_id(reaction_id) \
                .build()

            response = client.im.v1.message_reaction.delete(request)

            if not response.success():
                logger.warning('Failed to remove reaction', extra={
                    'message_id': message_id,
                    'reaction_id': reaction_id,
                    'code': response.code
                })

        except Exception as e:
            logger.warning('Failed to remove reaction', extra={
                'message_id': message_id,
                'error': str(e)
            })

    async def send_text_message(self, chat_id: str, text: str):
        """发送文本消息"""
        try:
            client = feishu_client.get_client()
            request = CreateMessageRequest.builder() \
                .receive_id_type('chat_id') \
                .request_body(CreateMessageRequestBody.builder()
                    .receive_id(chat_id)
                    .msg_type('text')
                    .content(json.dumps({'text': text}))
                    .build()) \
                .build()

            response = client.im.v1.message.create(request)

            if not response.success():
                logger.error('Failed to send text message', extra={
                    'chat_id': chat_id,
                    'code': response.code,
                    'msg': response.msg
                })
                raise Exception(f'发送消息失败: {response.msg}')

            logger.info('Text message sent', extra={'chat_id': chat_id})

        except Exception as e:
            logger.error('Failed to send text message', extra={
                'chat_id': chat_id,
                'error': str(e)
            })
            raise

    async def send_card_message(self, chat_id: str, markdown: str):
        """发送 Markdown 卡片消息"""
        try:
            client = feishu_client.get_client()

            card_content = {
                'config': {'wide_screen_mode': True},
                'elements': [
                    {'tag': 'markdown', 'content': markdown}
                ]
            }

            request = CreateMessageRequest.builder() \
                .receive_id_type('chat_id') \
                .request_body(CreateMessageRequestBody.builder()
                    .receive_id(chat_id)
                    .msg_type('interactive')
                    .content(json.dumps(card_content))
                    .build()) \
                .build()

            response = client.im.v1.message.create(request)

            if not response.success():
                logger.error('Failed to send card message', extra={
                    'chat_id': chat_id,
                    'code': response.code,
                    'msg': response.msg
                })
                raise Exception(f'发送卡片消息失败: {response.msg}')

            logger.info('Card message sent', extra={
                'chat_id': chat_id,
                'content_length': len(markdown)
            })

        except Exception as e:
            logger.error('Failed to send card message', extra={
                'chat_id': chat_id,
                'error': str(e)
            })
            raise


# 单例
message_service = MessageService()
